### PR TITLE
LLPC builder: matrix operations

### PR DESCRIPTION
**Pull request by trenouf**
(From JiaoLu)

Add builder of OpMatrixTimesScalar, OpVectorTimesMatrix,
OpMatrixTimesVector, OpMatrixTimesMatrix, OpOuterProduct, OpDotProduct.

Change-Id: I0e4677c6dc8e6b1362ba8b09bdb436367db9c2f5

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 5f17fdb6a1ba3caf4a79e633db71dbabfed3bc80
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev 46251728a7137f399b564b0886bd9bf9888c48ef</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev c740412cd4afe87d79a487bd375996d52360a3a0</li>
<li>LLVM: amd-vulkan-dev 9bc5dd4450a6361faf5c5661056a7ee494fad830</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 925/925 (100.0%)</li>
<li>Failed: 0/925 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 365531/365531 (100.0%)</li>
<li>Failed: 0/365531 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 365657/365657 (100.0%)</li>
<li>Failed: 0/365657 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by trenouf** [comment]
Thanks Qun, but it looks like you did not mark it approved.

File: builder/llpcBuilderImplMatrix.cpp [hunk 1]
```diff
@@ -89,3 +89,119 @@ Value* BuilderImplMatrix::CreateTransposeMatrix(
...
+
+// =====================================================================================================================
+// Create vector from vector Times matrix
+Value* BuilderImplMatrix::CreateVectorTimesMatrix(
+    Value* const pVector,         // [in] The float vector
+    Value* const pMatrix,         // [in] The column major matrix, n x <n x float>
+    const Twine& instName)        // [in] Name to give instruction(s)
+{
+    Type* const pMatrixTy = pMatrix->getType();
+    Type* const pElemTy = pMatrixTy->getArrayElementType()->getVectorElementType();
```
Comment (amdrexu) [inline builder/llpcBuilderImplMatrix.cpp]: pElemTy -> pCompTy

File: builder/llpcBuilderImplMatrix.cpp [hunk 2]
```diff
@@ -89,3 +89,119 @@ Value* BuilderImplMatrix::CreateTransposeMatrix(
...
+    Value* const pVector,             // [in] The vector
+    const Twine& instName)            // [in] Name to give instruction(s)
+{
+    Type* const pColumnTy = pMatrix->getType()->getArrayElementType();
+    const uint32_t rowCount = pColumnTy->getVectorNumElements();
+    Value* pResult = nullptr;
+
+    for (uint32_t i = 0; i < pMatrix->getType()->getArrayNumElements(); ++i)
+    {
+        SmallVector<uint32_t, 4> shuffleIndex(rowCount, i);
```
Comment (amdrexu) [inline builder/llpcBuilderImplMatrix.cpp]: shuffleIndex ->  shuffleMask

File: builder/llpcBuilderImplMatrix.cpp [hunk 3]
```diff
@@ -89,3 +89,119 @@ Value* BuilderImplMatrix::CreateTransposeMatrix(
...
+    const Twine& instName)            // [in] Name to give instruction(s)
+{
+    const uint32_t rowCount = pVector1->getType()->getVectorNumElements();
+    const uint32_t colCount = pVector2->getType()->getVectorNumElements();
+    Type* const pResultTy = ArrayType::get(pVector1->getType(), colCount);
+    Value* pResult = UndefValue::get(pResultTy);
+
+    for (uint32_t i = 0; i < colCount; ++i)
+    {
+        SmallVector<uint32_t, 4> shuffleIdx(rowCount, i);
```
Comment (amdrexu) [inline builder/llpcBuilderImplMatrix.cpp]: shuffleIdx ->  shuffleMask

File: builder/llpcBuilderRecorder.cpp [hunk 4]
```diff
@@ -229,6 +252,69 @@ Value* BuilderRecorder::CreateTransposeMatrix(
...
+
+// =====================================================================================================================
+// Create vector from vector times matrix
+Value* BuilderRecorder::CreateVectorTimesMatrix(
+    Value* const pVector,         // [in] The vector
+    Value* const pMatrix,         // [in] The matrix
+    const Twine& instName)        // [in] Name to give instruction(s)
+{
+    Type* const pMatrixType = pMatrix->getType();
+    Type* const pElemType = pMatrixType->getArrayElementType()->getVectorElementType();
```
Comment (amdrexu) [inline builder/llpcBuilderRecorder.cpp]: pElemType -> pCompType

File: builder/llpcBuilderRecorder.cpp [hunk 5]
```diff
@@ -229,6 +252,69 @@ Value* BuilderRecorder::CreateTransposeMatrix(
...
+
+// =====================================================================================================================
+// Create vector from matrix times vector
+Value* BuilderRecorder::CreateMatrixTimesVector(
+    Value* const pMatrix,             // [in] The matrix
+    Value* const pVector,             // [in] The vector
+    const Twine& instName)            // [in] Name to give instruction(s)
+{
+    Type* const pColumnType = pMatrix->getType()->getArrayElementType();
+    Type* const pElemType = pColumnType->getVectorElementType();
```
Comment (amdrexu) [inline builder/llpcBuilderRecorder.cpp]: pElemType -> pCompType

File: builder/llpcBuilderRecorder.cpp [hunk 6]
```diff
@@ -229,6 +252,69 @@ Value* BuilderRecorder::CreateTransposeMatrix(
...
+
+// =====================================================================================================================
+// Create matrix from outer product of vector
+Value* BuilderRecorder::CreateOuterProduct(
+    Value* const pVector1,            // [in] The vector 1
+    Value* const pVector2,            // [in] The vector 2
+    const Twine& instName)            // [in] Name to give instruction(s)
+{
+    const uint32_t colCount = pVector2->getType()->getVectorNumElements();
+    Type* const pMatrixType = ArrayType::get(pVector1->getType(), colCount);
```
Comment (amdrexu) [inline builder/llpcBuilderRecorder.cpp]: pMatrixType -> pResultTy

File: builder/llpcBuilderRecorder.cpp [hunk 7]
```diff
@@ -229,6 +252,69 @@ Value* BuilderRecorder::CreateTransposeMatrix(
...
+// =====================================================================================================================
+// Create matrix from matrix times matrix
+Value* BuilderRecorder::CreateMatrixTimesMatrix(
+    Value* const pMatrix1,             // [in] The matrix 1
+    Value* const pMatrix2,             // [in] The matrix 2
+    const Twine& instName)             // [in] Name to give instruction(s)
+{
+    Type* const pMat1ColumnType = pMatrix1->getType()->getArrayElementType();
+    const uint32_t mat2ColCount = pMatrix2->getType()->getArrayNumElements();
+    Type* const pMatrixType = ArrayType::get(pMat1ColumnType, mat2ColCount);
```
Comment (amdrexu) [inline builder/llpcBuilderRecorder.cpp]: pMatrixType -> pResultTy

File: builder/llpcBuilderRecorder.cpp [hunk 8]
```diff
@@ -229,6 +252,69 @@ Value* BuilderRecorder::CreateTransposeMatrix(
...
+// Create vector from matrix times vector
+Value* BuilderRecorder::CreateMatrixTimesVector(
+    Value* const pMatrix,             // [in] The matrix
+    Value* const pVector,             // [in] The vector
+    const Twine& instName)            // [in] Name to give instruction(s)
+{
+    Type* const pColumnType = pMatrix->getType()->getArrayElementType();
+    Type* const pElemType = pColumnType->getVectorElementType();
+    const uint32_t rowCount = pColumnType->getVectorNumElements();
+    Type* const pVectorType = VectorType::get(pElemType, rowCount);
```
Comment (amdrexu) [inline builder/llpcBuilderRecorder.cpp]: pVectorType -> pResultTy

File: builder/llpcBuilderRecorder.cpp [hunk 9]
```diff
@@ -229,6 +252,69 @@ Value* BuilderRecorder::CreateTransposeMatrix(
...
+// Create vector from vector times matrix
+Value* BuilderRecorder::CreateVectorTimesMatrix(
+    Value* const pVector,         // [in] The vector
+    Value* const pMatrix,         // [in] The matrix
+    const Twine& instName)        // [in] Name to give instruction(s)
+{
+    Type* const pMatrixType = pMatrix->getType();
+    Type* const pElemType = pMatrixType->getArrayElementType()->getVectorElementType();
+    const uint32_t columnCount = pMatrixType->getArrayNumElements();
+    Type* const pVectorType = VectorType::get(pElemType, columnCount);
```
Comment (amdrexu) [inline builder/llpcBuilderRecorder.cpp]: pVectorType -> pResultTy

**Comment by trenouf** [comment]
Thanks; I will land this.

**Comment by amdrexu** [comment]
> Thanks Qun, but it looks like you did not mark it approved.
Hi Tim,
Once a change is somewhat large, Qun or I will express the review ideas  and delay the approval until others have took a look.
BTW, I also see glslImageOpEmu.ll still exists in LLPC codes. I am not sure if it should be cleared as well.

**Comment by trenouf** [comment]
Thanks. Will check whether glslImageOpEmu.ll can now be removed.

**Comment by amdvlk-admin** [comment]
### Test summary for commit bc2d87dfe77c50e94616471d14a5f8dfeda27364
<details open><summary>Driver commits used in build</summary><ul>

<li>PAL: dev 46251728a7137f399b564b0886bd9bf9888c48ef</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev c740412cd4afe87d79a487bd375996d52360a3a0</li>
<li>LLVM: amd-vulkan-dev 9bc5dd4450a6361faf5c5661056a7ee494fad830</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 925/925 (100.0%)</li>
<li>Failed: 0/925 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 365531/365531 (100.0%)</li>
<li>Failed: 0/365531 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 365657/365657 (100.0%)</li>
<li>Failed: 0/365657 (0.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**